src/AppInstallerCLICore/ContextOrchestrator.cpp

```diff
@@ -145,13 +157,103 @@ namespace AppInstaller::CLI::Execution
         }
     }
 
-    _Requires_lock_held_(m_queueLock)
+    void ContextOrchestrator::RegisterForShutdownSynchronization()
+    {
+        static std::once_flag registerComponentOnceFlag;
+        std::call_once(registerComponentOnceFlag,
+            [&]()
+            {
+                using namespace ShutdownMonitoring;
+
+                ServerShutdownSynchronization::ComponentSystem component;
+                component.BlockNewWork = StaticDisable;
+                component.BeginShutdown = StaticCancelQueuedItems;
+                component.Wait = StaticWaitForRunningItems;
+
+                ServerShutdownSynchronization::AddComponent(component);
+            });
+    }
+
+    void ContextOrchestrator::StaticDisable(CancelReason reason)
+    {
+        Instance().Disable(reason);
+    }
+
+    void ContextOrchestrator::StaticCancelQueuedItems(CancelReason reason)
+    {
+        Instance().CancelQueuedItems(reason);
+    }
+
+    void ContextOrchestrator::StaticWaitForRunningItems()
+    {
+        Instance().WaitForRunningItems();
+    }
+
+    void ContextOrchestrator::Disable(CancelReason reason)
+    {
+        std::lock_guard<std::mutex> lock{ m_queueLock };
+        m_enabled = false;
+        m_disabledReason = reason;
+    }
+
+    void ContextOrchestrator::CancelQueuedItems(CancelReason reason)
+    {
+        std::lock_guard<std::mutex> lock{ m_queueLock };
+        for (const auto& queue : m_commandQueues)
+        {
+            queue.second->CancelAllItems(reason);
+        }
+    }
+
+    void ContextOrchestrator::WaitForRunningItems()
+    {
+        std::lock_guard<std::mutex> lock{ m_queueLock };
+        for (const auto& queue : m_commandQueues)
+        {
+            queue.second->WaitForEmptyQueue();
+        }
+    }
+
+    bool ContextOrchestrator::WaitForRunningItems(DWORD timeoutMilliseconds)
+    {
+        std::lock_guard<std::mutex> lock{ m_queueLock };
+        for (const auto& queue : m_commandQueues)
+        {
+            if (!queue.second->WaitForEmptyQueue(timeoutMilliseconds))

```

Comment:
Nit: I would expect this function to wait for _all_ items at most `timeoutMilliseconds`, but it waits that long for each of them. This may be moot depending on how many command queues we have or how big the timeout is, though.